src/mainwindow.cpp

```diff
@@ -1000,7 +1002,7 @@ void MainWindow::on_centralWidget_customContextMenuRequested(const QPoint &pos)
   createPopupMenu()->exec(ui->centralWidget->mapToGlobal(pos));
 }
 
-void MainWindow::scheduleUpdateButton()
+void MainWindow::scheduleCheckForProblems()
```

Comment:
Wouldn't the above code mean we are simply queuing all checks to be executed sequentially on the other thread? Like while a check is executed the main thread can set m_ChecksForProblemRequired again

Reply:
You're only queuing one "run". If the main thread set `m_ChecksForProblemRequired` while you're checking, yes, you will restart checking. But if the main thread set `m_ChecksForProblemRequired`  50 times while you're checking, you're not going to re-check 50 times, only once.

Reply:
I guess I would still want the 1 second buffer to collapse multiple calls even if they are fast.

Reply:
If the checks are really fast compared to the `checksForProblemAsync`  calls (e.g., you perform the check in 1ms and you call `checksForProblemAsync` every 2ms), then you're going to run it pretty often yes, but I don't think that's an issue here.

Reply:
Right now the problem is that we are potentially executing the checks at the same time since they are done asynchronously on a different thread so couldn't we just aquire a mutex in checkForContents ?

Reply:
You can keep the timer, but maybe make it much shorter (like 100ms)? You will anyway be limited (in the number of calls) by the speed of `checkForProblems`.

Reply:
You could acquire the mutex in `checkForContents` (should actually since some call do not go through the `Async` stuff), but in this case you could end up queuing a lot of calls if you call `Async` (without schedule) a lot. 

The idea of the threaded loop is that you're guaranteed to ever have a single call waiting in the "queue" (except for the direct to `checkForContents`).

Reply:
That would probably be a different mutex BTW, since one is to lock the call to `checkForProblems`  while the current one is to lock the "queue".

Reply:
Mostly I don't want to over-complicate with a dedicated thread, wait conditions etc since most of the time what is being changed here is irrelevant. I think I will add a mutex and we should just be aware of where we call the check synchronously. Right now we are doing it synchronously only when opening the dialog and even there I could make it async and use and event loop to wait for it but I don't think it's really necessary.

Reply:
I don't think that's over-complicated :) 

The issue is not only the synchronous call, it's also the async one. If you add a mutex, you're queuing all the async call on the mutex.

E.g., if you `checkForProblems`  takes 200ms, if you do 3 calls to `checkForProblemsAsync`  followed by one call to `checkForProblems` in less than 200ms, you will actually have to execute 4 times `checkForProblems`, and the last (sync) one will take 800ms because you'll have to wait for the 3 first ones to end.

With the loop version, you would only execute 2 calls to `checkForProblems`, so the synchronous one would only have to wait 200ms (the mutex) before executing.

Reply:
> Right now we are doing it synchronously only when opening the dialog and even there I could make it async and use and event loop to wait for it but I don't think it's really necessary.

That's actually a pretty good idea, that would remove the non-async `checkForProblems`  from the `MainWindow` interface, thus making it simpler.

Reply:
By the way, you probably don't need an extra thread for this, it might be possible to do everything with a `QFutureWatcher`  member that watches the asynchronous call.

You connect `finished()` to `checkForProblemsAsyncImpl` in which you check for a flag (so that the call does nothing if you have nothing to do). In `checkForProblemAsync`, you just set the flag to `true`  and call `checkForProblemsAsyncImpl`. You keep the timer `scheduleXXX`, just with a shorter timer.

Reply:
> You connect finished() to checkForProblemsAsyncImpl
This part is confusing me

Reply:
It's just that when the `QFuture` is done, you check if check was requested while running the future, and if yes, you start a new check. It similar to the `while (m_ChecksForProblemRequired.fetch_and(false))` in the loop-version.